### DISCUSSION TITLE
ModuleNotFoundError when running packed exe and how I solve it

### DISCUSSION THREAD

**Discussion by loktindyi** [pyinstaller/pyinstaller#7303] [Help] [answered]
(me: poor english and Amateur programmer, I'm sorry)
When I'm running an exe packed by pyinstaller, it thorw an error
> ModuleNotFoundError: No module named 'charset_normalizer.md__mypyc'
#### these infos might be useful:
I've imported these libs:
`openpyxl` `pdfplumber` `tqdm` `os` `sys`
I've installed 3.10 as 3.9 isnot uninstalled, these two versions have coexsited for a period, now 3.9 is uninstalled, I have plan to reinstall all environments
#### how I solved this problem
I tried to reinstall all libs, not work
I tried to pack it with arg. `-ondir`, then I found the dir of module`charset_normalizer` in the packed dir
I saw only one file named `md.cp310-win_amd64.pyd` in the `charset_normalizer` dir, which is not the exe need
I located to the module path `Lib\site-packages\charset_normalizer`, found these files in it:
1. `md.cp310-win_amd64.pyd`
2. `md.py`
3. `md__mypyc.cp310-win_amd64.pyd`

I realized that No.3 is what I need, there might be some problem when pyinstaller finding module, I guess
So I deleted file No.1, then repack and execute (Amateur practice)
It turns to normal

#### The question is
Why this problem appears and how to solve this normally, not like what I've done

**Comment by rokm** [comment, marked as answer]
> Why this problem appears and how to solve this normally, not like what I've done

This happens because `charset_normalizer.md__mypyc` is imported from `charset_normalizer.md`, which itself is a binary extension. PyInstaller cannot analyze the imports from binary extensions, and therefore `charset_normalizer.md__mypyc`  is not picked up automatically.

The correct way to solve this problem is to add the missing module(s) to *hidden imports* list, for example by adding [`--hiddenimport charset_normalizer.md__mypyc`](https://pyinstaller.org/en/stable/usage.html#cmdoption-hidden-import) to you PyInstaller command.

**Reply by loktindyi** [reply]
Thank you very much for your answer

**Reply by loktindyi** [reply]
By the way, do you know why pyinstaller can't pack on-demand libs just like webpack?
(I learn more about web, so this is an example I can give, haha.)
And will pyinstaller do support? 
Of course, it doesn't matter so much because we have venv, but it's always good to be more convenient.
